### PR TITLE
DOC: Correct documentation of ``__array__`` when used as output array.

### LINKED ISSUE #13458: DOC: Documentation suggest `__array__` interface is used for `out` kwarg to ufuncs
The documentation at: https://www.numpy.org/devdocs/reference/arrays.classes.html#numpy.class.__array__

suggests that when `out=non_ndarray` is used as kwarg to ufuncs, `non_ndarray.__array__` is called and used as output. This is not the case, and does not make much sense, since `__array__()` often/usually(?) can even return a copy.

The documentation should thus be fixed. Issue originally filed as pull request: gh-7318

### PR DESCRIPTION
**Pull request by tinaoberoi**
- Update `__array__` interface behaviour in documentation
- Add update file in docs/release/upcoming_changes

Closes https://github.com/numpy/numpy/issues/13458

### CODE REVIEW

File: doc/release/upcoming_changes/13458.changes.rst [hunk 1]
```diff
@@ -0,0 +1,4 @@
+Update the documentation for `__array__` interface behaviour 
+--------------------------------------------------
+Earlier the `__array__` documentation said:
+If a class (ndarray subclass or not) having the `__array__` method is used as the output object of an ufunc, results will be written to the object returned by `__array__` However the behaviour of `__array__` did not support it, the documentation was modified to avoid confusion.
```
Comment (eric-wieser) [inline doc/release/upcoming_changes/13458.changes.rst:4]: Thanks for trying to be thorough - but there's no need for release notes for documentation changes.

Comment (tinaoberoi) [reply doc/release/upcoming_changes/13458.changes.rst]: Thank you @eric-wieser. I have made the suggested changes.

**Review by rgommers — COMMENTED** [review]
Thanks for working on this @tinaoberoi. Your content change looks good, the formatting has a tiny issue with backticks:
- ``` ``not`` ``` double backticks means "typeset as code", and I think you want italics (so ``*not*``).
- ``` `TypeError` ``` single backticks means "create link to this reference" (and that isn't found, so it ends up in italics), what you want is typeset as code (so ``` ``TypeError`` ```).

Could you change that please?

**Comment by rossbar** [comment]
Re: the content of the proposed change - as I commented in the corresponding issue, I wonder if the doc for `__array__` should focus more on what it does generally, rather than it's behavior in one specific condition, i.e. when used as the input with the `out` parameter. Maybe this behavior is documented elsewhere, but I feel that a high-level description of `__array__` makes more sense in the section on special attrs/methods than something so specific.

The current doc (and the proposed change) don't really fit with the descriptions of the other special methods, which describe what they are for more generally. I think this approach should be taken here as well.

**Comment by rgommers** [comment]
@rossbar your observation is very valid, and if @tinaoberoi would like to work (either in this PR or a new one) on that that's great. I'm happy to merge this PR as is though, since it fixes a reported doc inaccuracy. We too often put too many extra requirements on PRs, which keeps them from being merged for too long. A separate follow-up issue instead is better.

**Comment by tinaoberoi** [comment]
> Thanks for working on this @tinaoberoi. Your content change looks good, the formatting has a tiny issue with backticks:
> 
> * `` ``not`` `` double backticks means "typeset as code", and I think you want italics (so `*not*`).
> * `` `TypeError` `` single backticks means "create link to this reference" (and that isn't found, so it ends up in italics), what you want is typeset as code (so `` ``TypeError`` ``).
> 
> Could you change that please?

Thanks for pointing that out @rgommers I have made the necessary changes.

**Comment by rgommers** [comment]
Great, LGTM now - merged, thanks @tinaoberoi

**Pull request merged**